tests_integration/nativeapp/test_deploy.py

```diff
@@ -421,3 +421,103 @@ def test_nativeapp_deploy_path_with_no_mapping(
+
+    with pushd(Path(os.getcwd(), project_dir)):
+        try:
+            result = runner.invoke_with_connection_json(
+                ["app", "deploy", "-r", "app"],
+                env=TEST_ENV,
+            )
+            assert result.exit_code == 0
+
+            package_name = f"{project_name}_pkg_{USER_NAME}".upper()
+            stage_name = "app_src.stage"  # as defined in native-apps-templates/basic
+            stage_files = runner.invoke_with_connection_json(
+                ["stage", "list-files", f"{package_name}.{stage_name}"],
+                env=TEST_ENV,
+            )
+            assert contains_row_with(stage_files.json, {"name": "stage/manifest.yml"})
+            assert contains_row_with(
+                stage_files.json, {"name": "stage/setup_script.sql"}
+            )
+            assert contains_row_with(stage_files.json, {"name": "stage/README.md"})
+
+        finally:
+            result = runner.invoke_with_connection_json(
+                ["app", "teardown", "--force"],
+                env=TEST_ENV,
+            )
+            assert result.exit_code == 0
+
+
+# Tests that snow app deploy . -r --prune is the same as the default deploy command
+@pytest.mark.integration
+def test_nativeapp_deploy_dot(
+    runner,
+    temporary_working_directory,
+):
+    project_name = "myapp"
+    project_dir = "app root"
+    result = runner.invoke_json(
+        ["app", "init", project_dir, "--name", project_name],
+        env=TEST_ENV,
+    )
+    assert result.exit_code == 0
+
+    with pushd(Path(os.getcwd(), project_dir)):
+        try:
+            result = runner.invoke_with_connection_json(
+                ["app", "deploy", "-r", "."],
+                env=TEST_ENV,
+            )
+            assert result.exit_code == 0
+
+            package_name = f"{project_name}_pkg_{USER_NAME}".upper()
+            stage_name = "app_src.stage"  # as defined in native-apps-templates/basic
+            stage_files = runner.invoke_with_connection_json(
+                ["stage", "list-files", f"{package_name}.{stage_name}"],
+                env=TEST_ENV,
+            )
+            assert contains_row_with(stage_files.json, {"name": "stage/manifest.yml"})
+            assert contains_row_with(
+                stage_files.json, {"name": "stage/setup_script.sql"}
+            )
+            assert contains_row_with(stage_files.json, {"name": "stage/README.md"})
+
+            os.remove(os.path.join("app", "README.md"))
+
+            # deploy
+            result = runner.invoke_with_connection_json(
+                ["app", "deploy", "-r", "--prune", "."],
+                env=TEST_ENV,
+            )
+            assert result.exit_code == 0
+
+            stage_files = runner.invoke_with_connection_json(
+                ["stage", "list-files", f"{package_name}.{stage_name}"],
+                env=TEST_ENV,
+            )
+
+            assert contains_row_with(stage_files.json, {"name": "stage/README.md"})
```

Comment:
I'm confused why this works, considering you just removed the local source of `stage/README.md` and called deploy with `--prune`. Can you explain it?

Reply:
Great catch, I really thought I had written the flipped condition here, which would indeed make the test fail. What's happening is that prune + file arguments doesn't really work. Once a file is deleted, we lose track of why it was on the stage in the first place. We see that stage/README.md is present, but we filter it out because we have no idea where it came from. I see three possible ways to handle that:

1. Disallow `--prune` when arguments are passed in to deploy. It doesn't work intuitively, it will only prune things that match the local path anyway. At least we'd disallow it until we fix the behaviour, likely by doing some bookkeeping locally to remember mappings for deleted files.
2. Have `--prune` always prune _all_ matches, regardless of the arguments provided. That's potentially surprising to users, since `deploy -r --prune app` doesn't limit the pruning to the specified directory.
3. Have `--prune` work as it does now, which is that if you map artifacts to the same directory in the deploy root (e.g. src -> src), then prune will work as you expect, otherwise it will have no effect (as shown here, since app -> .)

In all cases, I think we need to fix the behaviour. We've talked about the long term solution before in the context of auto-reload, we have a design for it that should be revisited. In the meantime though, I prefer solution 1, but I'm open to arguments for the other options as well. Thoughts?

Reply:
The first approach makes the most sense to me. Let's make the command error out when `--prune` and files/folders are given, and then update the documentation:

- if we are saying `snow app deploy` is essentially an alias for `snow app deploy --prune -r .` we'll need to drop that `.`
- document the fact that one dest folder can be populated from multiple source folders, which is why `--prune` cannot work with files

Reply:
Agreed, the alias thing would now become aspirational, when we fix the prune logic. Will make the necessary changes here and resubmit then, I think it makes sense to bundle all the changes in one PR.